python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+
+    # Method to do things needed to actively running tests.
+    # NOTE: We are in a thread locked state, so be quick
+    def handleActiveTests(self):
+        # loop through all the active tests and do something
+        for tester_name in self.active_jobs:
+            tester = self.scheduled_name_to_object[tester_name]
+
+            # Handle long running tests
+            self.doLongRunningTests(tester)
+
+            # Handle timeouts
+            self.doTimeoutTests(tester)
+
+    def doLongRunningTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            min_time = self.default_reported_time
+            if tester.specs.isValid('min_reported_time'):
+                min_time = float(tester.getMinReportTime())
+
+            if clock() - tester.getStartTime() > min_time:
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('RUNNING...', tester.bucket_pending)
+                self.harness.handleTestStatus(tester)
+
+    def doTimeoutTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            if clock() - tester.getStartTime() > float(tester.getMaxTime()):
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('TIMEOUT', tester.bucket_fail)
+                self.harness.handleTestStatus(tester)
+
+    # TODO: Possibly look into semaphore functionality for slot allocation
+    def checkAvailableSlots(self, tester):
+        # See if we can fit this job into our busy schedule
+        if self.workers_in_use + tester.getProcs(self.options) <= self.job_slots:
+            return True
+
+        # Check for insufficient slots -soft limit
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and self.soft_limit:
+            tester.specs.addParam('caveats', ['OVERSIZED'], "")
+            return True
+
+        # Check for insufficient slots -hard limit (skip this job)
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and not self.soft_limit:
+            tester.setStatus('insufficient slots', tester.bucket_skip)
+
+    # Status processing (hand a tester's status back to the TestHarness)
+    def statusRunner(self, queue):
+        tester_name = queue.get()
+        tester = self.scheduled_name_to_object[tester_name]
+
+        # Test is Finished (failed, passed, skipped, silent, etc)
+        if tester.isFinished():
+            # And the running caveat
+            if tester_name in self.reported_jobs and tester.didPass():
+                tester.specs.addParam('caveats', ['FINISHED'], "")
+
+            # Print finalized results
+            self.harness.handleTestStatus(tester)
+
+            # Remove this finished tester from the DAG and active job set, get the next node, and assign it to
+            # a runner thread
+            with self.thread_lock:
```

Comment:
Don't need the lock here. You only need it where you start modifying the dag. Keep the locks as close to the operations they are protecting as possible.